hive-metastore/src/test/java/org/apache/iceberg/hive/TestCachedClientPool.java

```diff
@@ -46,6 +46,10 @@ public void testClientPoolCleaner() throws InterruptedException {
     Assert.assertNull(
         CachedClientPool.clientPoolCache()
             .getIfPresent(CachedClientPool.extractKey(null, hiveConf)));
+
+    // The client has been really closed.
+    Assert.assertTrue(clientPool1.isClosed());
```

Comment:
Just for the record: Did we test that this test fails without the fix?

Reply:
I've checked this when reviewing and the test fails with the code as-is. But the test passes with 
```
Awaitility.await()
        .atMost(10, TimeUnit.SECONDS)
        .untilAsserted(
            () -> {
              Assert.assertTrue(clientPool1.isClosed());
              Assert.assertTrue(clientPool2.isClosed());
            });
```
because cache removal will eventually be executed (just not immediately)

Reply:
Caffeine's doc describes its complicated cleanup behavior, and introduces the scheduler to prompt cleanup.
 
`By default, Caffeine does not perform cleanup and evict values "automatically" or instantly after a value expires. Instead, it performs small amounts of maintenance work after write operations or occasionally after read operations if writes are rare. `
( see: https://github.com/ben-manes/caffeine/wiki/Cleanup )

The unpredictability of removal could increase the pressure on HMS, when there are many applications (including Flink and Spark).